WIP: Documentation and Examples (#257)
OutWatch/outwatch: README.md

Source Code: [DomTypes.scala](src/main/scala/outwatch/dom/DomTypes.scala)

#### Data attributes
Data attributes make use of [`scala.Dynamic`](https://www.scala-lang.org/api/current/scala/Dynamic.html), so you can write things like:

```scala
div(data.payload := "17")
// <div data-payload="17"></div>
```

Source Code: [OutwatchAttributes.scala](src/main/scala/outwatch/dom/OutwatchAttributes.scala#L70), [Builder.scala](src/main/scala/outwatch/dom/helpers/Builder.scala#L39)

#### SVG
SVG tags and attributes are available via an extra import. Namespacing is automatically handled for you.

```scala
val graphic = {
 import svg._
 svg(
   viewBox := "0 0 10 10",
   g(
     transform := "matrix(.096584 0 0 .096584 -.0071925 -18.66)",
     path(d := "M10 10 C 20 20, 40 20, 50 10", fill := "mistyrose")
   )
 )
}

// <svg viewBox="0 0 10 10"><g transform="matrix(.096584 0 0 .096584 -.0071925 -18.66)"><path d="M10 10 C 20 20, 40 20, 50 10" fill="mistyrose"></path></g></svg>
```

**TODO**: Make an svg which actually shows something.

Custom SVG attributes and tags are possible like this:

```scala

> TODO: svgTag? svgAttr?

Reviewer: Should we remove this if it's not possible yet? :)

Author: Ah, yes, I will move this into the documentation issue.

    
    val myComponent = div("Hello World")

    OutWatch.renderReplace("#app", myComponent).unsafeRunSync()
  }
}
```

Running `Main` will replace `<div id="app"></div>` with the content defined in `myComponent`:

```html
...
<body>
    <div>Hello World</div>
    ...
</body>
...
```

### Static Content
First, we will focus on creating immutable/static content that will not change over time. The following examples illustrate to construct and transform HTML/SVG tags, attributes and inline stylesheets.

#### Concatenating Strings
```scala
div("Hello ", "World")
// <div>Hello World</div>
```

#### Nesting
```scala
div(span("Hey ", b("you"), "!"))
// <div>Hey <b>you</b>!</div>
```

#### Primitives
```scala
div(true, 0, 1000L, 3.0)
// <div>true010003.0</div>
```

#### Attributes
Attributes are put inside the tag.

```scala
div(id := "test")
// <div id="test"></div>
```

The order of content and attributes does not matter.

```scala
div("How ", id := "test", "are", title := "cool", " you?")
// <div id="test" title="cool">How are you?</div>
```

#### Styles
Styles are also written into the tag. All style properties have to be written in *camelCase*.

```scala

> div(color := "tomato", "Hello")

Reviewer 2: ```suggestion
div(color := "blue", "Hello")
```

Author: why not tomato? :grin:

Reviewer 2: Is there a CSS color named 🍅? 😁

Author: yes! It's a nice red color :ok_hand:: https://www.quackit.com/html/codes/color/color-tester.cfm?colorname=tomato

> // <div style="color:blue">Hello</div>

Reviewer 2: ```suggestion
// <div style="color: blue">Hello</div>
```

```

Multiple styles will me merged to one style attribute:

```scala
div(backgroundColor := "powderblue", border := "2px solid #222", "Hello")

> // <div style="background-color:powderblue; border: 2px solid #222">Hello</div>

Reviewer 2: ```suggestion
// <div style="background-color: powderblue; border: 2px solid #222">Hello</div>
```

```

Again, the order of styles, attributes and inner tags does not matter:

```scala
div(h1("Welcome to my website"), backgroundColor := "powderblue", id := "header")

> // <div style="background-color:powderblue" id="header">Welcome to my website</div>

Reviewer 2: ```suggestion
// <div style="background-color: powderblue" id="header">Welcome to my website</div>
```

// <label for="inputid" />
```

There is also a shortcut for the class atrribute:

```scala
div(cls := "myclass")
// <div class="myclass"></div>
```

Source Code: [OutwatchAttributes.scala](src/main/scala/outwatch/dom/OutwatchAttributes.scala#L65)

#### Overriding attributes
Attributes and styles with the same name will be overwritten. Last wins.

```scala
div(color := "blue", color := "green")
// <div style="color: green"></div>
```

Source Code: [DomUtils.scala](src/main/scala/outwatch/dom/helpers/DomUtils.scala#L8)

#### Class accumulation
Classes are not overwritten, they accumulate.

```scala
div(cls := "tiny", cls := "button")
// <div class="tiny button"></div>
```

#### Custom attributes, styles and tags
All the tags, attributes and styles available in outwatch come from [Scala Dom Types](https://github.com/raquo/scala-dom-types).
If you want to use something not available in Scala Dom Types, you can use custom builders:

```scala
htmlTag("app")(style("user-select") := "none", attr("everything") := "possible")
// <app style="user-select: none" everything="possible"></div>
```

You can also define the accumulation behavior of custom attributes:
```scala
div(

>   attr("everything").accum("-") := "im",

Reviewer 2: ```suggestion
  attr("everything").accum("-") := "is",
```

  attr("everything").accum("-") := "possible",
)

> // <div everything="im-possible"></div>

Reviewer 2: ```suggestion
// <div everything="is-possible"></div>
```

     path(d := "M10 10 C 20 20, 40 20, 50 10", fill := "mistyrose")
   )
 )
}

// <svg viewBox="0 0 10 10"><g transform="matrix(.096584 0 0 .096584 -.0071925 -18.66)"><path d="M10 10 C 20 20, 40 20, 50 10" fill="mistyrose"></path></g></svg>
```

**TODO**: Make an svg which actually shows something.

Custom SVG attributes and tags are possible like this:

```scala
TODO: svgTag? svgAttr?
```

#### Option and Seq
Outwatch tries hard to render everything you throw at it. Combine `Option` and `Seq` to fit your needs. Note, that outwatch does not accept `Set`, since the order is undefined.

```scala
div(
  Some("thing"),
  Some(color := "steelblue"),
  fontSize :=? Some("70px"),
  None,
  Seq("Hey", "How are you?"),
  List("a", "b", "c").map(span(_)),
  Some(Seq("x")),
)
// <div style="color: steelblue; font-size: 70px;">
//   thing
//   Hey
//   How are you?
//   <span>a</span>
//   <span>b</span>
//   <span>c</span>
//   x
// </div>
```

Source Code: [AsVDomModifier.scala](src/main/scala/outwatch/AsVDomModifier.scala)

#### Types
The important types we were using in the examples above are `VNode` and `VDomModifier`:

```scala

> val vnode:VNode = div()

Reviewer 2: ```suggestion
val vnode: VNode = div()
```

> val modifiers:List[VDomModifier] = List("Hello", id := "main", color := "tomato", vnode)

Reviewer 2: ```suggestion
val modifiers: List[VDomModifier] = List("Hello", id := "main", color := "tomato", vnode)
```

```

Every `VNode` contains a sequence of `VDomModifier`. A `VNode` is a `VDomModifier` itself.

There are implicits for converting primitives, `Option[VDomModifier]`, `Seq[VDomModifier]` to `VDomModifier`.

#### Grouping Modifiers
To make a set of modifiers reusable you can group them to become one `VDomModifier`.

```scala
val bigFont = VDomModifier(fontSize := "40px", fontWeight.bold)
div("Argh!", bigFont)
// <div style="font-size: 40px; font-weight: bold;">Argh!</div>
```

> Of course you could also use a `Seq[VDomModifier]` instead.

Reviewer 2: ```suggestion
You can also use a `Seq[VDomModifier]` directly instead of using `apply` defined in the [VDomModifier](src/main/scala/outwatch/dom/package.scala) object.
```

```

Source Code: [VDomModifier.scala](src/main/scala/outwatch/dom/VDomModifier.scala#L146)

#### Use-Case: Flexbox
When working with [Flexbox](https://css-tricks.com/snippets/css/a-guide-to-flexbox/), you can set styles for the **container** and **children**. With `VNode.apply()` you can have all flexbox-related styles in one place. The child-components don't have to know anything about flexbox, even though they get specific styles assigned.

```scala
val itemA = div("A", backgroundColor := "mediumseagreen")
val itemB = div("B", backgroundColor := "tomato")

div(
  height := "100px",
  border := "1px solid black",

  display.flex,

  itemA(flexBasis := "50px"),
  itemB(alignSelf.center),
)
// <div style="height: 100px; border: 1px solid black; display: flex;">
//   <div style="background-color: mediumseagreen; flex-basis: 50px;">A</div>
//   <div style="background-color: tomato; align-self: center;">B</div>
// </div>
```

### Dynamic Content
To visualize updates, use an `Observable[VDomModifier]` as if it was a `VDomModifier`.

```scala
import monix.reactive.Observable
import concurrent.duration._
    
val counter = Observable.interval(1 second)
div("count: ", counter)
```

> 

Reviewer 2: Would be great to have live examples of this inside some box :)

Author: You are right. There should at least be a link to a jsfiddle.

Reviewer: Maybe we could think of using something like `sbt-microsite` it's pretty easy to set up :)

Author: Yes, even better with a combination of compiling examples like https://github.com/scalameta/mdoc. You can add that to the documentation issue, to address this in a future change.
